saleor/payment/gateways/razorpay/__init__.py

```diff
@@ -132,26 +134,23 @@ def refund(payment: Payment, amount: Decimal, **connection_params):
     payment id to refund. And return an error with a failed transaction
     if the there is no such transaction, or if an error
     from razorpay occurs during the refund."""
-    error = check_payment_supported(payment=payment)
-    capture_txn = payment.transactions.filter(
-        kind=TransactionKind.CHARGE, is_success=True).first()
+    error = check_payment_supported(payment_information=payment_information)
 
     if error:
-        response = get_error_response(amount)
-    elif capture_txn is not None:
+        response = get_error_response(
+            payment_information['amount'], error=error)
+    else:
         razorpay_client = get_client(**connection_params)
-        razorpay_amount = get_amount_for_razorpay(amount)
+        razorpay_amount = get_amount_for_razorpay(
+            payment_information['amount'])
         try:
             response = razorpay_client.payment.refund(
-                capture_txn.token, razorpay_amount)
+                payment_information['token'], razorpay_amount)
             clean_razorpay_response(response)
         except RAZORPAY_EXCEPTIONS as exc:
             error = get_error_message_from_razorpay_error(exc)
-            response = get_error_response(amount)
-    else:
-        error = errors.ORDER_NOT_CHARGED
-        response = get_error_response(amount)
+            response = get_error_response(
+                payment_information['amount'], error=error)
 
-    transaction = _generate_transaction(
-        payment=payment, kind=TransactionKind.REFUND, **response)
-    return transaction, error
+    return _generate_response(
+        payment_information=payment_information, kind='refund', **response)
```

Comment:
We shouldn't be setting the transaction kinds through hard-coded strings, that may make it hard to refactor, or at least to maintain Saleor.

Reply:
There is a validation in ```validate_gateway_response``` by ```if response['kind'] not in ALLOWED_GATEWAY_KINDS:```. But it is better if it is not hard-coded.

Reply:
The problem is that eventually those gateways won't live in Saleor, so they won't have access to the `TrasnactionKind` that Saleor has, so there are two ways of moving forward:
1) make a common package that implements such things
2) just hard code the strings
I took the second approach for the time being, we can always move to the other one if necessary.

Reply:
I think what @NyanKiyoshi means though these transaction types won't be avaible out of saleor, it still be nice to wrapper it as an enum or whatever in each gateways. So in each gateway, it doesn't need to access the hard-coded string such ```refund``` multiple time. It only needs to access the field of that transaction kind enum. It makes it easier to refactor.

Reply:
Ah fair, sounds good, I can do that.